src/dlfcn.c

```diff
@@ -467,7 +472,162 @@ char *dlerror( void )
     return error_buffer;
 }
 
-#ifdef SHARED
+/* taken from http://bandido.ch/programming/Import_Address_Table_Hooking.pdf */
+IMAGE_IMPORT_DESCRIPTOR* getImportTable( HMODULE module, DWORD *size )
+{
+    IMAGE_DOS_HEADER* dosHeader = (IMAGE_DOS_HEADER*)module;
+    if ( dosHeader->e_magic != 0x5A4D )
+        return NULL;
+    IMAGE_OPTIONAL_HEADER* optionalHeader = (IMAGE_OPTIONAL_HEADER*)
+        ((BYTE*)module + dosHeader->e_lfanew + 24);
+    if ( optionalHeader->Magic != IMAGE_NT_OPTIONAL_HDR_MAGIC )
+        return NULL;
+    if ( optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size == 0 ||
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return (IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress);
+}
+
+#ifdef _WIN64
+typedef ULONGLONG PointerType;
+#else
+typedef ULONG PointerType;
+#endif
+
+/*
+ * return symbol name for a given address
+ */
+char *getSymbolName( HMODULE baseAddress, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    PointerType base = (PointerType)baseAddress;
+    for(int i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
+        PIMAGE_THUNK_DATA thunkILT = (PIMAGE_THUNK_DATA)(iid[i].Characteristics + base);
+        PIMAGE_THUNK_DATA thunkIAT = (PIMAGE_THUNK_DATA)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if (thunkIAT->u1.Function != (PointerType)addr)
+              continue;
+            PIMAGE_IMPORT_BY_NAME nameData = (PIMAGE_IMPORT_BY_NAME)(thunkILT->u1.AddressOfData + base);
+            return nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static unsigned char *getAddressFromIAT( unsigned char *addr )

```

Comment:
Both return value and function parameter should be generic `void *` pointer type. Both input and output is not string (char *).

Reply:
> function parameter should be generic void * pointer type. Both input and output is not string (char*)
Then 
```
   if( addr[0] != 0xff || addr[1] != 0x25 )
        return NULL;
```
fails to compile 
```
home2/ralf/src/dlfcn-win32/src/dlfcn.c:527:13: warning: dereferencing 'void *' pointer
  527 |     if( addr[0] != 0xff || addr[1] != 0x25 )
      |             ^
/home2/ralf/src/dlfcn-win32/src/dlfcn.c:527:13: error: void value not ignored as it ought to be
  527 |     if( addr[0] != 0xff || addr[1] != 0x25 )
      |         ~~~~^~~
/home2/ralf/src/dlfcn-win32/src/dlfcn.c:527:32: warning: dereferencing 'void *' pointer
  527 |     if( addr[0] != 0xff || addr[1] != 0x25 )
      |                                ^
/home2/ralf/src/dlfcn-win32/src/dlfcn.c:527:32: error: void value not ignored as it ought to be
  527 |     if( addr[0] != 0xff || addr[1] != 0x25 )
```
That was the reason I used a BYTE *p pointer

Reply:
But it is not reason to "leak" char* type (string) outside of the whole function and force caller to pass string into this function when input is not string. You should do explicit casts if they are needed, e.g.:

```c
static void* getAddressFromIAT(void *addr)
{
    if( *(short *)addr == 0x25ff )
        return NULL;
}
```

Reply:
You have not fixed return value. It is still `char *`.